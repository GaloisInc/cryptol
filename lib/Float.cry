module Float where

primitive type ValidFloat : # -> # -> Prop

/** IEEE-754 floating point numbers. */
primitive type { exponent : #, precision : #}
  ValidFloat exponent precision => Float exponent precision : *

/** An abbreviation for common 16-bit floating point numbers. */
type Float16  = Float 5 11

/** An abbreviation for common 32-bit floating point numbers. */
type Float32  = Float 8 24

/** An abbreviation for common 64-bit floating point numbers. */
type Float64  = Float 11 53

/** An abbreviation for common 128-bit floating point numbers. */
type Float128 = Float 15 113

/** An abbreviation for common 256-bit floating point numbers. */
type Float256 = Float 19 237



/* ----------------------------------------------------------------------
 * Rounding modes (this should be an enumeration type, when we add these)
 *---------------------------------------------------------------------- */

type RoundingMode = [3]

/** Round toward nearest, ties go to even. */
roundNearestEven, rne : RoundingMode
roundNearestEven = 0
rne              = roundNearestEven

/** Round toward nearest, ties away from zero. */
roundNearestAway, rna : RoundingMode
roundNearestAway  = 1
rna               = roundNearestAway

/** Round toward positive infinity. */
roundPositive, rtp : RoundingMode
roundPositive     = 2
rtp               = roundPositive

/** Round toward negative infinity. */
roundNegative, rtn : RoundingMode
roundNegative     = 3
rtn               = roundNegative

/** Round toward zero. */
roundZero, rtz : RoundingMode
roundZero         = 4
rtz               = roundZero



/* ----------------------------------------------------------------------
 * Constants
 * ---------------------------------------------------------------------- */

/** Not a number. */
primitive
  fpNaN : {e,p} ValidFloat e p => Float e p

/** Positive infinity. */
primitive
  fpPosInf : {e,p} ValidFloat e p => Float e p

fpNegInf : {e,p} ValidFloat e p => Float e p
fpNegInf = - fpPosInf

/** Positive zero. */
fpPosZero : {e,p} ValidFloat e p => Float e p
fpPosZero = zero

/** Negative zero. */
fpNegZero : {e,p} ValidFloat e p => Float e p
fpNegZero = - fpPosZero


// Binary representations

/** A floating point number using the exact bit pattern,
in IEEE interchange format with layout:

  (sign : [1]) # (biased_exponent : [e]) # (significand : [p-1])
*/
primitive
  fpFromBits : {e,p} ValidFloat e p => [e + p] -> Float e p

/** Export a floating point number in IEEE interchange format with layout:

  (sign : [1]) # (biased_exponent : [e]) # (significand : [p-1])

NaN is represented as:
  * positive:           sign        == 0
  * quiet with no info: significand == 0b1 # 0
*/
primitive
  fpToBits : {e,p} ValidFloat e p => Float e p -> [e + p]





/* ----------------------------------------------------------------------
 * Predicates
 * ----------------------------------------------------------------------
 */

// Operations in `Cmp` use IEEE reasoning.

/** Check if two floating point numbers are representationally the same.
In particular, the following hold:
    *    NaN       =.= NaN
    * ~ (pfNegZero =.= fpPosZero)
*/
primitive
  (=.=) : {e,p} ValidFloat e p => Float e p -> Float e p -> Bool

infix 20 =.=




/* ----------------------------------------------------------------------
 * Arithmetic
 * ---------------------------------------------------------------------- */


/** Add floating point numbers using the given rounding mode. */
primitive
  fpAdd : {e,p} ValidFloat e p =>
    RoundingMode -> Float e p -> Float e p -> Float e p

/** Subtract floating point numbers using the given rounding mode. */
primitive
  fpSub : {e,p} ValidFloat e p =>
    RoundingMode -> Float e p -> Float e p -> Float e p

/** Multiply floating point numbers using the given rounding mode. */
primitive
  fpMul : {e,p} ValidFloat e p =>
    RoundingMode -> Float e p -> Float e p -> Float e p

/** Divide floating point numbers using the given rounding mode. */
primitive
  fpDiv : {e,p} ValidFloat e p =>
    RoundingMode -> Float e p -> Float e p -> Float e p




