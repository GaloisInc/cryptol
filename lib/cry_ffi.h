#pragma once

#include <stdlib.h>
#include <stdint.h>

struct CryValueBuilder {
  /** This should be passed to all the function pointers.  */
  void*     self;

  /** Make a boolean value: 0: false, > 0: true */
  void      (*send_bool)(void*, uint8_t);

  /** Make an unsigned Integer, which fits in uint64_t */
  void      (*send_small_uint)(void*, uint64_t);

  /** Make an signed Integer, which fits in int64_t */
  void      (*send_small_sint)(void*, int64_t);

  /** Start building a sum type,
      the argument is the number of the constructor */
  void      (*send_tag)(void*, size_t);

  /** Start building an Integer, which does not fit in int64_t.
      The argument is the number of uint64_t digits we need.
      The result is a buffer where the digits should be placed,
      least significant first.
   */
  uint64_t* (*new_large_int)(void*, size_t);

  /** Finish bulding an Integer, which does not fit in uint64_t.
      The argument is 1 for negative, and 0 for non-negative.
   */
  void      (*send_sign)(void*, uint8_t);
};