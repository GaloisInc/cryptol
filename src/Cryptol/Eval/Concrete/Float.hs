{-# Language BlockArguments #-}
{-# Language OverloadedStrings #-}
{-# Language BangPatterns #-}
-- | Concrete evaluations for floating point primitives.
module Cryptol.Eval.Concrete.Float where

import Data.Map(Map)
import Data.Bits(testBit,setBit,shiftL,shiftR,(.&.),(.|.))
import Data.Int(Int64)
import qualified Data.Map as Map
import LibBF

import Cryptol.Utils.Ident(PrimIdent, floatPrim)
import Cryptol.Utils.Panic(panic)
import Cryptol.Eval.Value
import Cryptol.Eval.Generic
import Cryptol.Eval.Concrete.Value
import Cryptol.Eval.Concrete.FloatHelpers



floatPrims :: Concrete -> Map PrimIdent Value
floatPrims sym = Map.fromList [ (floatPrim i,v) | (i,v) <- nonInfixTable ]
  where
  (~>) = (,)
  nonInfixTable =
    [ "fpNaN"       ~> ilam \e -> ilam \p ->
                        VFloat BF { bfValue = bfNaN
                                  , bfExpWidth = e, bfPrecWidth = p }

    , "fpPosInf"    ~> ilam \e -> ilam \p ->
                       VFloat BF { bfValue = bfPosInf
                                 , bfExpWidth = e, bfPrecWidth = p }

    , "fpFromBits"  ~> ilam \e -> ilam \p -> wlam sym \bv ->
                       pure $ VFloat $ floatFromBits e p $ bvVal bv

    , "fpToBits"    ~> ilam \e -> ilam \p -> flam \x ->
                       pure $ word sym (e + p)
                            $ floatToBits e p
                            $ bfValue x
    , "=.="         ~> ilam \_ -> ilam \_ -> flam \x -> pure $ flam \y ->
                       pure $ VBit
                            $ bitLit sym
                            $ bfCompare (bfValue x) (bfValue y) == EQ

      -- From Backend class
    , "fpAdd"      ~> fpBinArithV sym fpPlus
    , "fpSub"      ~> fpBinArithV sym fpMinus
    , "fpMul"      ~> fpBinArithV sym fpMult
    , "fpDiv"      ~> fpBinArithV sym fpDiv
    ]


floatFromBits :: 
  Integer {- ^ Exponent width -} ->
  Integer {- ^ Precision widht -} ->
  Integer {- ^ Raw bits -} ->
  BF
floatFromBits e p bv = BF { bfValue = floatFromBits' e p bv
                          , bfExpWidth = e, bfPrecWidth = p }



-- | Make a float using "raw" bits.
floatFromBits' ::
  Integer {- ^ Exponent width -} ->
  Integer {- ^ Precision widht -} ->
  Integer {- ^ Raw bits -} ->
  BigFloat

floatFromBits' e p bits
  | expoBiased == 0 && mant == 0 =            -- zero
    if isNeg then bfNegZero else bfPosZero

  | expoBiased == eMask && mant ==  0 =       -- infinity
    if isNeg then bfNegInf else bfPosInf

  | expoBiased == eMask = bfNaN               -- NaN

  | expoBiased == 0 =                         -- Subnormal
    case bfMul2Exp opts (bfFromInteger mant) (expoVal + 1) of
      (num,Ok) -> if isNeg then bfNeg num else num
      (_,s)    -> panic "floatFromBits" [ "Unexpected status: " ++ show s ]

  | otherwise =                               -- Normal
    case bfMul2Exp opts (bfFromInteger mantVal) expoVal of
      (num,Ok) -> if isNeg then bfNeg num else num
      (_,s)    -> panic "floatFromBits" [ "Unexpected status: " ++ show s ]

  where
  opts       = expBits e' <> precBits (p' + 1) <> allowSubnormal

  e'         = fromInteger e                               :: Int
  p'         = fromInteger p - 1                           :: Int
  eMask      = (1 `shiftL` e') - 1                         :: Int64
  pMask      = (1 `shiftL` p') - 1                         :: Integer

  isNeg      = testBit bits (e' + p')

  mant       = pMask .&. bits                              :: Integer
  mantVal    = mant `setBit` p'                            :: Integer
  -- accounts for the implicit 1 bit

  expoBiased = eMask .&. fromInteger (bits `shiftR` p')    :: Int64
  bias       = eMask `shiftR` 1                            :: Int64
  expoVal    = expoBiased - bias - fromIntegral p'         :: Int64


-- | Turn a float into raw bits.
-- @NaN@ is represented as a positive "quiet" @NaN@
-- (most significant bit in the significand is set, the rest of it is 0)
floatToBits :: Integer -> Integer -> BigFloat -> Integer
floatToBits e p bf =  (isNeg      `shiftL` (e' + p'))
                  .|. (expBiased  `shiftL` p')
                  .|. (mant       `shiftL` 0)
  where
  e' = fromInteger e     :: Int
  p' = fromInteger p - 1 :: Int

  eMask = (1 `shiftL` e') - 1   :: Integer
  pMask = (1 `shiftL` p') - 1   :: Integer

  (isNeg, expBiased, mant) =
    case bfToRep bf of
      BFNaN       -> (0,  eMask, 1 `shiftL` (p' - 1))
      BFRep s num -> (sign, be, ma)
        where
        sign = case s of
                Neg -> 1
                Pos -> 0

        (be,ma) =
          case num of
            Zero     -> (0,0)
            Num i ev
              | ex == 0   -> (0, i `shiftL` (p' - m  -1))
              | otherwise -> (ex, (i `shiftL` (p' - m)) .&. pMask)
              where
              m    = msb 0 i - 1
              bias = eMask `shiftR` 1
              ex   = toInteger ev + bias + toInteger m

            Inf -> (eMask,0)

  msb !n j = if j == 0 then n else msb (n+1) (j `shiftR` 1)

