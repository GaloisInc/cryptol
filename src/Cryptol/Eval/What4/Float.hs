{-# Language BlockArguments #-}
{-# Language OverloadedStrings #-}
-- | Floating point primitives for the What4 backend.
module Cryptol.Eval.What4.Float (floatPrims) where

import Data.Map(Map)
import qualified Data.Map as Map
import qualified What4.Interface as W4
import Control.Monad.IO.Class

import Cryptol.TypeCheck.Solver.InfNat(Nat'(..))
import Cryptol.Eval.Value
import Cryptol.Eval.Generic
import Cryptol.Eval.What4.Value
import qualified Cryptol.Eval.What4.SFloat as W4
import Cryptol.Utils.Ident(PrimIdent, floatPrim)

-- | Table of floating point primitives
floatPrims :: W4.IsExprBuilder sym => What4 sym -> Map PrimIdent (Value sym)
floatPrims sym4@(What4 sym) =
  Map.fromList [ (floatPrim i,v) | (i,v) <- nonInfixTable ]
  where
  (~>) = (,)

  nonInfixTable =
    [ "fpNaN"       ~> fpConst (W4.fpNaN sym)
    , "fpPosInf"    ~> fpConst (W4.fpPosInf sym)
    , "fpFromBits"  ~> ilam \e -> ilam \p -> wlam sym4 \w ->
                       VFloat <$> liftIO (W4.fpFromBinary sym e p w)
    , "fpToBits"    ~> ilam \e -> ilam \p -> flam \x ->
                       pure $ VWord (e+p)
                            $ WordVal <$> liftIO (W4.fpToBinary sym x)
    , "=.="         ~> ilam \_ -> ilam \_ -> flam \x -> pure $ flam \y ->
                       VBit <$> liftIO (W4.fpEq sym x y)

    , "fpAdd"       ~> fpBinArithV sym4 fpPlus
    , "fpSub"       ~> fpBinArithV sym4 fpMinus
    , "fpMul"       ~> fpBinArithV sym4 fpMult
    , "fpDiv"       ~> fpBinArithV sym4 fpDiv
    ]



-- | A helper for definitng floating point constants.
fpConst ::
  W4.IsExprBuilder sym =>
  (Integer -> Integer -> IO (W4.SFloat sym)) ->
  Value sym
fpConst mk =
     ilam \ e ->
 VNumPoly \ ~(Nat p) ->
 VFloat <$> liftIO (mk e p)


