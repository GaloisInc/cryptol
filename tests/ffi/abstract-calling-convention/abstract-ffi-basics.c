#include <assert.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cry_ffi.h"

void test_bool(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t v;
    assert(CRY_FFI(args,recv_u8,&v) == 0);
    CRY_FFI(res,send_bool,v);
}

void test_u8(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t v;
    assert(CRY_FFI(args,recv_u8,&v) == 0);
    CRY_FFI(res,send_small_uint,(uint64_t)(v));
}

void test_u16(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint64_t v;
    assert(CRY_FFI(args,recv_u64,&v) == 0);
    CRY_FFI(res,send_small_uint,v);
}

void test_u32(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_u16(args, res);
}

void test_u64(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_u16(args, res);
}

void test_i8(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t v;
    assert(CRY_FFI(args,recv_u8,&v) == 0);
    CRY_FFI(res,send_small_sint,(int64_t)(v));
}

void test_i16(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint64_t v;
    assert(CRY_FFI(args,recv_u64,&v) == 0);
    CRY_FFI(res,send_small_sint,(int64_t)(v));
}

void test_i32(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_i16(args, res);
}

void test_i64(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_i16(args, res);
}

static uint64_t* recv_integer(uint8_t* sign, uint64_t* size, const struct CryValExporter* args) {
    assert(CRY_FFI(args,recv_u8,sign) == 0);
    assert(CRY_FFI(args,recv_u64,size) == 0);
    uint64_t* digits = calloc(*size, sizeof(uint64_t));
    assert(CRY_FFI(args,recv_u64_digits,digits) == 0);
    return digits;
}

static void send_integer(uint8_t sign, uint64_t size, const uint64_t* recv_digits, const struct CryValImporter* res) {
    uint64_t* send_digits = CRY_FFI(res,send_new_large_int,size);
    memcpy(send_digits, recv_digits, size);
    CRY_FFI(res,send_sign,sign);
}

void test_integer(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t sign;
    uint64_t size;
    uint64_t* digits = recv_integer(&sign, &size, args);
    send_integer(sign, size, digits, res);
    free(digits);
}

void test_rational(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t numer_sign;
    uint64_t numer_size;
    uint64_t* numer_digits = recv_integer(&numer_sign, &numer_size, args);

    uint8_t denom_sign;
    uint64_t denom_size;
    uint64_t* denom_digits = recv_integer(&denom_sign, &denom_size, args);

    send_integer(numer_sign, numer_size, numer_digits, res);
    send_integer(denom_sign, denom_size, denom_digits, res);

    free(numer_digits);
    free(denom_digits);
}

void test_float(const struct CryValExporter* args, const struct CryValImporter* res) {
    double v;
    assert(CRY_FFI(args,recv_double,&v) == 0);
    CRY_FFI(res,send_double,v);
}

void test_double(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_float(args, res);
}

void test_array_16_u8(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t v[16];
    for (int i = 0; i < 16; i++) {
        assert(CRY_FFI(args,recv_u8,&v[i]) == 0);
    }
    for (int i = 0; i < 16; i++) {
        CRY_FFI(res,send_small_uint,(uint64_t)(v[i]));
    }
}

void test_record0(const struct CryValExporter* args, const struct CryValImporter* res) {}

void test_record1_u8(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_u8(args, res);
}

void test_record2_u8_u16(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t v0;
    uint64_t v1;

    assert(CRY_FFI(args,recv_u8,&v0) == 0);
    assert(CRY_FFI(args,recv_u64,&v1) == 0);

    CRY_FFI(res,send_small_uint,(uint64_t)(v0));
    CRY_FFI(res,send_small_uint,v1);
}

void test_tuple0(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_record0(args, res);
}

void test_tuple2_u8_u16(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_record2_u8_u16(args, res);
}

void test_newtype_u8(const struct CryValExporter* args, const struct CryValImporter* res) {
    test_record1_u8(args, res);
}

static void unexpected_tag(const char* enum_name, uint64_t tag) {
    printf("Unexpected tag for %s: %" PRIu64 "\n", enum_name, tag);
    abort();
}

void test_option_u8(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint64_t tag;
    uint8_t field;
    assert(CRY_FFI(args,recv_u64,&tag) == 0);
    switch (tag) {
    case 0:
        // None
        break;
    case 1:
        // Some
        assert(CRY_FFI(args,recv_u8,&field) == 0);
        break;
    default:
        unexpected_tag("Option", tag);
        break;
    }

    CRY_FFI(res,send_tag,tag);
    switch (tag) {
    case 0:
        // None
        break;
    case 1:
        // Some
        CRY_FFI(res,send_small_uint,(uint64_t)(field));
        break;
    default:
        unexpected_tag("Option", tag);
        break;
    }
}

void test_result_u16_u32(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint64_t tag;
    uint64_t field;
    assert(CRY_FFI(args,recv_u64,&tag) == 0);
    switch (tag) {
    case 0:
        // Ok
        assert(CRY_FFI(args,recv_u64,&field) == 0);
        break;
    case 1:
        // Err
        assert(CRY_FFI(args,recv_u64,&field) == 0);
        break;
    default:
        unexpected_tag("Result", tag);
        break;
    }

    CRY_FFI(res,send_tag,tag);
    switch (tag) {
    case 0:
        // Ok
        CRY_FFI(res,send_small_uint,field);
        break;
    case 1:
        // Err
        CRY_FFI(res,send_small_uint,field);
        break;
    default:
        unexpected_tag("Result", tag);
        break;
    }
}

void test_two_u8_args(const struct CryValExporter* args, const struct CryValImporter* res) {
    uint8_t v0;
    uint8_t v1;

    assert(CRY_FFI(args,recv_u8,&v0) == 0);
    assert(CRY_FFI(args,recv_u8,&v1) == 0);

    CRY_FFI(res,send_small_uint,(uint64_t)(v0 - v1));
}
