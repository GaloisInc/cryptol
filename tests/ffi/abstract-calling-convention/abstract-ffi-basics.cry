import Float

newtype N = { unN : [8] }

foreign abstract test_bool : Bit -> Bit
foreign abstract test_u8 : [8] -> [8]
foreign abstract test_u16 : [16] -> [16]
foreign abstract test_u32 : [32] -> [32]
foreign abstract test_u64 : [64] -> [64]
foreign abstract test_u128 : [128] -> [128]
foreign abstract test_i8 : [8] -> [8]
foreign abstract test_i16 : [16] -> [16]
foreign abstract test_i32 : [32] -> [32]
foreign abstract test_i64 : [64] -> [64]
foreign abstract test_i128 : [128] -> [128]
foreign abstract test_integer : Integer -> Integer
foreign abstract test_rational : Rational -> Rational
foreign abstract test_float : Float32 -> Float32
foreign abstract test_double : Float64 -> Float64
foreign abstract test_array_16_u8 : [16][8] -> [16][8]
foreign abstract test_record0 : {} -> {}
foreign abstract test_record1_u8 : { x : [8] } -> { x : [8] }
foreign abstract test_record2_u8_u16 : { x : [8], y : [16] } -> { x : [8], y : [16] }
foreign abstract test_tuple0 : () -> ()
foreign abstract test_tuple2_u8_u16 : ([8], [16]) -> ([8], [16])
foreign abstract test_newtype_u8 : N -> N
foreign abstract test_option_u8 : Option [8] -> Option [8]
foreign abstract test_result_u16_u32 : Result [16] [32] -> Result [16] [32]
foreign abstract test_two_u8_args : [8] -> [8] -> [8]
foreign abstract test_two_u128_res : [128] -> ([128], [128])
foreign abstract test_z7 : Z 7 -> Z 7
foreign abstract test_param : {n} (fin n, n >= 1) => Z n
