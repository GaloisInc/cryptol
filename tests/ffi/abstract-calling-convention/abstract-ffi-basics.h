#include <cry_ffi.h>
void test_bool(const struct CryValExporter * args, const struct CryValImporter * res);
void test_u8(const struct CryValExporter * args, const struct CryValImporter * res);
void test_u16(const struct CryValExporter * args, const struct CryValImporter * res);
void test_u32(const struct CryValExporter * args, const struct CryValImporter * res);
void test_u64(const struct CryValExporter * args, const struct CryValImporter * res);
void test_i8(const struct CryValExporter * args, const struct CryValImporter * res);
void test_i16(const struct CryValExporter * args, const struct CryValImporter * res);
void test_i32(const struct CryValExporter * args, const struct CryValImporter * res);
void test_i64(const struct CryValExporter * args, const struct CryValImporter * res);
void test_integer(const struct CryValExporter * args, const struct CryValImporter * res);
void test_rational(const struct CryValExporter * args, const struct CryValImporter * res);
void test_float(const struct CryValExporter * args, const struct CryValImporter * res);
void test_double(const struct CryValExporter * args, const struct CryValImporter * res);
void test_array_16_u8(const struct CryValExporter * args, const struct CryValImporter * res);
void test_record0(const struct CryValExporter * args, const struct CryValImporter * res);
void test_record1_u8(const struct CryValExporter * args, const struct CryValImporter * res);
void test_record2_u8_u16(const struct CryValExporter * args, const struct CryValImporter * res);
void test_tuple0(const struct CryValExporter * args, const struct CryValImporter * res);
void test_tuple2_u8_u16(const struct CryValExporter * args, const struct CryValImporter * res);
void test_newtype_u8(const struct CryValExporter * args, const struct CryValImporter * res);
void test_option_u8(const struct CryValExporter * args, const struct CryValImporter * res);
void test_result_u16_u32(const struct CryValExporter * args, const struct CryValImporter * res);
void test_two_u8_args(const struct CryValExporter * args, const struct CryValImporter * res);
void test_z7(const struct CryValExporter * args, const struct CryValImporter * res);
void test_param(const struct CryValExporter * args, const struct CryValImporter * res);
