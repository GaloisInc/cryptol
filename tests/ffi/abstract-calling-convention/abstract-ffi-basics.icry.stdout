Loading module Cryptol
Loading module Cryptol
Loading module Float
Loading module Main
Loading dynamic library abstract-ffi-basics.so
False
0x01
0x0001
0x00000001
0x0000000000000001
0x01
0x0001
0x00000001
0x0000000000000001
42
-42
(ratio 2 5)
0x1.8
0x1.8
[0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a,
 0x0b, 0x0c, 0x0d, 0x0e, 0x0f]
{}
{x = 0x00}
{x = 0x00, y = 0x0001}
()
(0x00, 0x0001)
{unN = 0x01}
None
Some 0x01
Ok 0x0001
Err 0x00000002
0x03
